oauthenticator/github.py

```diff
@@ -215,6 +239,31 @@ async def authenticate(self, handler, data=None):
                     auth_state['github_user']['email'] = val['email']
                     break
 
+        if self.fetch_teams:
+            # Number of teams to request per page
+            per_page = 100
+            # Page counter
+            cur_page = 1
+
+            teams = []
+            while True:
+                req = HTTPRequest(
+                    self.github_api
+                    + f"/user/teams?per_page={per_page}&page={cur_page}",
+                    method="GET",
+                    headers=_api_headers(access_token),
+                    validate_cert=self.validate_server_cert,
+                )
+                resp_json = await self.fetch(req, "fetching user teams")
+                teams += resp_json
+
+                # Only try to paginate if we received the max number of teams we could have
```

Comment:
Does the teams API response include pagination information?
https://docs.github.com/en/github-ae@latest/rest/overview/resources-in-the-rest-api#pagination

Reply:
I don't see it. If I set `per_page` to `1` to trigger this easy, here's first response for me:

```
[{'name': 'Owners', 'id': 66678, 'node_id': 'MDQ6VGVhbTY2Njc4', 'slug': 'owners', 'description': None, 'privacy': 'secret', 'url': 'https://api.github.com/organizations/867692/team/66678', 'html_url': 'https://github.com/orgs/publictransit-in/teams/owners', 'members_url': 'https://api.github.com/organizations/867692/team/66678/members{/member}', 'repositories_url': 'https://api.github.com/organizations/867692/team/66678/repos', 'permission': 'admin', 'created_at': '2007-10-20T11:24:19Z', 'updated_at': '2007-10-20T11:24:19Z', 'members_count': 4, 'repos_count': 7, 'organization': {'login': 'publictransit-in', 'id': 867692, 'node_id': 'MDEyOk9yZ2FuaXphdGlvbjg2NzY5Mg==', 'url': 'https://api.github.com/orgs/publictransit-in', 'repos_url': 'https://api.github.com/orgs/publictransit-in/repos', 'events_url': 'https://api.github.com/orgs/publictransit-in/events', 'hooks_url': 'https://api.github.com/orgs/publictransit-in/hooks', 'issues_url': 'https://api.github.com/orgs/publictransit-in/issues', 'members_url': 'https://api.github.com/orgs/publictransit-in/members{/member}', 'public_members_url': 'https://api.github.com/orgs/publictransit-in/public_members{/member}', 'avatar_url': 'https://avatars.githubusercontent.com/u/867692?v=4', 'description': 'Community projects around public transport in India', 'name': 'India public transport tools', 'company': None, 'blog': 'http://publictransit.in', 'location': 'India', 'email': 'arun.planemad@gmail.com', 'twitter_username': None, 'is_verified': False, 'has_organization_projects': True, 'has_repository_projects': True, 'public_repos': 8, 'public_gists': 0, 'followers': 0, 'following': 0, 'html_url': 'https://github.com/publictransit-in', 'created_at': '2011-06-22T17:54:15Z', 'updated_at': '2016-01-26T20:18:57Z', 'type': 'Organization'}, 'parent': None}]
```

Just a list without anything else.

Reply:
Anything in the response headers? The docs imply there should be a `Link:`

Reply:
I'm linking here the docs that mention the Link header as the location of pagination info: https://docs.github.com/en/github-ae@latest/rest/guides/traversing-with-pagination#basics-of-pagination

Reply:
Ah, I didn't inspect it but makes sense it would be in the response header.

I'm slightly torn on what to do here. I can look for `Link` in the header and try use that, but that means I can't use the nice `self.fetch` function. But that's probably the 'right' thing to do, especially given the specific warning about it in the GItHub API docs...

Should I move the code to stop using `self.fetch` and use a http client directly?

Reply:
It's more work, but code has a habit of being copied and pasted so I think it's better to do it properly. I suspect it'll also be useful in future if others extend this class to fetch additional information in their own deployments.

Reply:
Makes sense.. I'll do that.

Reply:
@manics reworked to use the Link header now!